server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorDriverMetadata.java

```diff
@@ -21,41 +21,96 @@
 
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
 
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public class AppenderatorDriverMetadata
 {
-  private final Map<String, List<SegmentIdentifier>> activeSegments;
-  private final Map<String, List<SegmentIdentifier>> publishPendingSegments;
+  private final Map<String, List<AppenderatorDriver.SegmentWithState>> segments;
   private final Map<String, String> lastSegmentIds;
   private final Object callerMetadata;
 
   @JsonCreator
   public AppenderatorDriverMetadata(
-      @JsonProperty("activeSegments") Map<String, List<SegmentIdentifier>> activeSegments,
-      @JsonProperty("publishPendingSegments") Map<String, List<SegmentIdentifier>> publishPendingSegments,
+      @JsonProperty("segments") Map<String, List<AppenderatorDriver.SegmentWithState>> segments,
       @JsonProperty("lastSegmentIds") Map<String, String> lastSegmentIds,
-      @JsonProperty("callerMetadata") Object callerMetadata
+      @JsonProperty("callerMetadata") Object callerMetadata,
+      // Next two properties are for backwards compatibility, should be removed on versions greater than 0.11.1
+      @JsonProperty("activeSegments") Map<String, List<SegmentIdentifier>> activeSegments,
+      @JsonProperty("publishPendingSegments") Map<String, List<SegmentIdentifier>> publishPendingSegments
   )
   {
-    this.activeSegments = activeSegments;
-    this.publishPendingSegments = publishPendingSegments;
+    Preconditions.checkState(
+        segments != null || (activeSegments != null && publishPendingSegments != null),
+        "Metadata should either have segments with state information or both active segments and publish pending segments information. "
+        + "segments [%s], activeSegments [%s], publishPendingSegments [%s]",
+        segments,
+        activeSegments,
+        publishPendingSegments
+    );
+    if (segments == null) {
```

Comment:
@pjain1 would you let me know when ```segments``` can be null? 
Also, ```activeSegments``` and ```publishPendingSegments``` look always null. Maybe these should be removed but accidentally remained?

Reply:
Oh, never mind. I just found your comment on these variables.

Reply:
This is done for backwards compatibility so that old AppenderatorDriverMetadata can be converted to new format. `segments` will be null while reading older metadata before upgrading the middle manager code, after upgrade the old metadata needs to be converted to new one.

Yes, for new code `activeSegments` and `publishPendingSegments` will always be null therefore these properties can be removed on versions greater than 0.12.x

Reply:
Right. Thanks.